### PR TITLE
[Telemetry] Add function `NumOfBorderRoutersAddedExternalRoute()`

### PR DESCRIPTION
**Pull request by zesonzhang**
This will be used for future divergent BR detection

### CODE REVIEW

**Comment by bukepo** [comment]
Could you add some a test?

File: src/utils/thread_helper.cpp [hunk 1]
```diff
@@ -1048,6 +1048,28 @@ void ThreadHelper::RetrievePdProcessedRaInfo(threadnetwork::TelemetryData::PdPro
...
 #endif // OTBR_ENABLE_DHCP6_PD
 
+uint8_t ThreadHelper::NumberOfBrPublishedExternalRoute()
+{
+    std::set<uint16_t> rloc16Set;
+
+    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
+    otExternalRouteConfig config;
+
+    while (otNetDataGetNextRoute(mInstance, &iterator, &config) == OT_ERROR_NONE)
```
[the remark this answers is not in the document]
Comment (librasungirl) [reply src/utils/thread_helper.cpp:1058]: sounds good to ignore nat64 as offline discussion the counters here are mainly for evaluating AIL.

File: src/utils/thread_helper.cpp [hunk 2]
```diff
@@ -1048,6 +1048,28 @@ void ThreadHelper::RetrievePdProcessedRaInfo(threadnetwork::TelemetryData::PdPro
...
+    {
+        if (!config.mStable)
+        {
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.cpp:1063]: My understanding is that when counting the total number of the BRs in the partition, we expect all the BRs to be "stable". Also, I saw only stable entries are able to be published (https://github.com/openthread/openthread/blob/4c0d8f2e5992c40dbb92829245bd35d68674a9bc/include/openthread/netdata_publisher.h#L220). Any thoughts?

File: src/utils/thread_helper.cpp [hunk 3]
```diff
@@ -1048,6 +1048,28 @@ void ThreadHelper::RetrievePdProcessedRaInfo(threadnetwork::TelemetryData::PdPro
...
+
+        rloc16Set.insert(config.mRloc16);
+    }
+
+    // the size is no more than OPENTHREAD_CONFIG_NETDATA_PUBLISHER_DESIRED_NUM_EXTERNAL_ROUTE_ENTRIES
+    // it should be safe to convert to uint8_t
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.cpp:1069]: Thanks for the clarification!

File: src/utils/thread_helper.hpp [hunk 4]
```diff
@@ -332,6 +332,12 @@ class ThreadHelper
     void RetrieveHashedPdPrefix(std::string *aHashedPdPrefix);
     void RetrievePdProcessedRaInfo(threadnetwork::TelemetryData::PdProcessedRaInfo *aPdProcessedRaInfo);
 #endif
+    /**
+     * @brief Retrieve network data and counts the total number of BRs that have published external routes.
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.hpp:336]: oops... have mixed "published" and "added" for many times  🤣  Thanks for reminding me again!

File: src/utils/thread_helper.cpp [hunk 5]
```diff
@@ -1048,6 +1048,28 @@ void ThreadHelper::RetrievePdProcessedRaInfo(threadnetwork::TelemetryData::PdPro
 }
 #endif // OTBR_ENABLE_DHCP6_PD
 
+uint8_t ThreadHelper::NumberOfBrPublishedExternalRoute()
+{
+    std::set<uint16_t> rloc16Set;
+
+    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.cpp:1055]: Thanks Abtin! Looks like `CountBorderRouters()` is perfectly what I needed. I pushed https://github.com/openthread/openthread/pull/10487, and this PR should be no longer needed

**Comment by zesonzhang** [comment]
https://github.com/openthread/openthread/pull/10487 is the alternative solution

File: src/utils/thread_helper.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.cpp]: updated the description

File: src/utils/thread_helper.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.hpp]: updated the description

File: src/utils/thread_helper.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (abtink) [reply src/utils/thread_helper.cpp:1055]: Thanks. I wish I posted earlier. I submitted this PR yesterday.  It adds mechanism to track peer BRs found Network Data and also tracks their age, i.e., how long they have been present in the network data. It also provides a public API to get the total count. 
- https://github.com/openthread/openthread/pull/10485

[the remark this answers is not in the document]
Comment (zesonzhang) [reply src/utils/thread_helper.cpp:1055]: Thanks! #10485 looks great! I will close this one

**Comment by zesonzhang** [comment]
This feature in implemented in https://github.com/openthread/openthread/pull/10485

**Pull request closed without merging**